virtio-devices/src/vdpa.rs

```diff
@@ -0,0 +1,374 @@
+                log_addr: None,
+            };
+
+            self.vhost
+                .set_vring_addr(queue_index, &config_data)
+                .map_err(Error::SetVringAddr)?;
+            self.vhost
+                .set_vring_base(
+                    queue_index,
+                    queue
+                        .avail_idx(Ordering::Acquire)
+                        .map_err(Error::GetAvailableIndex)?
+                        .0,
+                )
+                .map_err(Error::SetVringBase)?;
+
+            if let Some(eventfd) =
+                virtio_interrupt.notifier(VirtioInterruptType::Queue(queue_index as u16))
+            {
+                self.vhost
+                    .set_vring_call(queue_index, &eventfd)
+                    .map_err(Error::SetVringCall)?;
+            }
+
+            self.vhost
+                .set_vring_kick(queue_index, &queue_evts[queue_index])
+                .map_err(Error::SetVringKick)?;
+        }
+
+        // Setup the config eventfd if there is one
+        if let Some(eventfd) = virtio_interrupt.notifier(VirtioInterruptType::Config) {
+            self.vhost
+                .set_config_call(&eventfd)
+                .map_err(Error::SetConfigCall)?;
+        }
+
+        self.enable_vrings(queues.len(), true)?;
+
+        self.vhost
+            .set_status(
+                (DEVICE_ACKNOWLEDGE | DEVICE_DRIVER | DEVICE_DRIVER_OK | DEVICE_FEATURES_OK) as u8,
+            )
+            .map_err(Error::SetStatus)
+    }
+
+    fn reset_vdpa(&mut self) -> Result<()> {
+        if let Some(num_queues) = self.enabled_num_queues {
+            self.enable_vrings(num_queues, false)?;
+        }
+
+        self.vhost.set_status(0).map_err(Error::SetStatus)
+    }
+
+    fn dma_map(&self, iova: u64, size: u64, host_vaddr: *const u8, readonly: bool) -> Result<()> {
+        if iova < self.iova_range.first || iova + size - 1 > self.iova_range.last {
+            return Err(Error::InvalidIovaRange);
+        }
+
+        self.vhost
+            .dma_map(iova, size, host_vaddr, readonly)
+            .map_err(Error::DmaMap)
+    }
+
+    fn dma_unmap(&self, iova: u64, size: u64) -> Result<()> {
+        if iova < self.iova_range.first || iova + size - 1 > self.iova_range.last {
+            return Err(Error::InvalidIovaRange);
+        }
+
+        self.vhost.dma_unmap(iova, size).map_err(Error::DmaUnmap)
+    }
+}
+
+impl VirtioDevice for Vdpa {
+    fn device_type(&self) -> u32 {
+        self.common.device_type
+    }
+
+    fn queue_max_sizes(&self) -> &[u16] {
+        &self.common.queue_sizes
+    }
+
+    fn features(&self) -> u64 {
+        self.common.avail_features
+    }
+
+    fn ack_features(&mut self, value: u64) {
+        self.common.ack_features(value)
+    }
+
+    fn read_config(&self, offset: u64, data: &mut [u8]) {
+        if let Err(e) = self.vhost.get_config(offset as u32, data) {
+            error!("vdpa: Failed reading virtio config: {}", e);
```

Comment:
`error!()` include source file and line so i'm not sure the extra "vdpa" prefix adds anything?